lib/Interpreter/Transaction.cpp

```diff
@@ -62,12 +62,22 @@ namespace cling {
   }
 
   NamedDecl* Transaction::containsNamedDecl(llvm::StringRef name) const {
-    for (auto I = decls_begin(), E = decls_end(); I < E; ++I)
+    for (auto I = decls_begin(), E = decls_end(); I < E; ++I) {
       for (auto DI : I->m_DGR) {
-        if (NamedDecl* ND = dyn_cast<NamedDecl>(DI))
+        if (NamedDecl* ND = dyn_cast<NamedDecl>(DI)) {
           if (name.equals(ND->getNameAsString()))
             return ND;
+        }
+        else if (LinkageSpecDecl* LSD = dyn_cast<LinkageSpecDecl>(DI)) {
```

Comment:
I am thinking how to make this interface work in all possible cases. One of the problems is that we can find only first level declarations, which makes the interface really awkward. We will need some sort of visitor but this makes the implementation complicated. IIRC, the only use is in MetaSema for diagnostic purposes which can be replaced with a normal C++ lookup. Then we can ask the Transaction whether it contains the found Decl* (which is a far more simple job).

Another approach would be to simplify and speed up its implementation of Transaction::containsNamedDecl by performing a SemaLookup and check if the found Decl* were in the range of the transaction. We know the range of the Decl* in a transaction because of the bump allocator used for allocation. Eg. if the pointer is between the pointer of Transaction's first declaration and last declaration that would mean that this transaction contains the Decl (something like clang::DeclContext::containsDecl).